internal/provider/kubernetes/predicates.go

```diff
@@ -516,6 +517,142 @@ func (r *gatewayAPIReconciler) validateEndpointSliceForReconcile(obj client.Obje
+		return false
+	}
+	for _, route := range httpRouteList.Items {
+		if r.hasEndpointRouting(route.Namespace, route.Spec.CommonRouteSpec) {
+			return true
+		}
+	}
+
+	if r.grpcRouteCRDExists {
+		grpcRouteList := &gwapiv1.GRPCRouteList{}
+		if err := r.client.List(ctx, grpcRouteList, &client.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector(backendGRPCRouteIndex, nsName.String()),
+		}); err != nil && !kerrors.IsNotFound(err) {
+			r.log.Error(err, "failed to find associated GRPCRoutes")
+			return false
+		}
+		for _, route := range grpcRouteList.Items {
+			if r.hasEndpointRouting(route.Namespace, route.Spec.CommonRouteSpec) {
+				return true
+			}
+		}
+	}
+
+	if r.tlsRouteCRDExists {
+		tlsRouteList := &gwapiv1a2.TLSRouteList{}
+		if err := r.client.List(ctx, tlsRouteList, &client.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector(backendTLSRouteIndex, nsName.String()),
+		}); err != nil && !kerrors.IsNotFound(err) {
+			r.log.Error(err, "failed to find associated TLSRoutes")
+			return false
+		}
+		for _, route := range tlsRouteList.Items {
+			if r.hasEndpointRouting(route.Namespace, route.Spec.CommonRouteSpec) {
+				return true
+			}
+		}
+	}
+
+	if r.tcpRouteCRDExists {
+		tcpRouteList := &gwapiv1a2.TCPRouteList{}
+		if err := r.client.List(ctx, tcpRouteList, &client.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector(backendTCPRouteIndex, nsName.String()),
+		}); err != nil && !kerrors.IsNotFound(err) {
+			r.log.Error(err, "failed to find associated TCPRoutes")
+			return false
+		}
+		for _, route := range tcpRouteList.Items {
+			if r.hasEndpointRouting(route.Namespace, route.Spec.CommonRouteSpec) {
+				return true
+			}
+		}
+	}
+
+	if r.udpRouteCRDExists {
+		udpRouteList := &gwapiv1a2.UDPRouteList{}
+		if err := r.client.List(ctx, udpRouteList, &client.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector(backendUDPRouteIndex, nsName.String()),
+		}); err != nil && !kerrors.IsNotFound(err) {
+			r.log.Error(err, "failed to find associated UDPRoutes")
+			return false
+		}
+		for _, route := range udpRouteList.Items {
+			if r.hasEndpointRouting(route.Namespace, route.Spec.CommonRouteSpec) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// hasEndpointRouting checks that the associated egv1a1.EnvoyProxy has endpoint routing.
+func (r *gatewayAPIReconciler) hasEndpointRouting(namespace string, spec gwapiv1.CommonRouteSpec) bool {
+	ctx := context.Background()
+	for _, ref := range spec.ParentRefs {
+		if ref.Kind != nil && *ref.Kind != resource.KindGateway {
+			return false
+		}
+		if ref.Namespace != nil {
+			namespace = string(*ref.Namespace)
+		}
+
+		gw := gwapiv1.Gateway{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      string(ref.Name),
+				Namespace: namespace,
+			},
+		}
+		err := r.client.Get(ctx, client.ObjectKeyFromObject(&gw), &gw)
+		if err != nil {
+			r.log.Error(err, "unable to find associated gateway")
+			return false
+		}
+
```

Comment:
the `Gateway` resource can also have a `infrastructure.parameteresRef` field, which takes precedence over the one in GatewayClass, so if thats not nil, we should use that `EnvoyProxy` as source of truth